src/box/lua/trigger.c

```diff
@@ -253,3 +253,146 @@ box_lua_trigger_init(struct lua_State *L)
+{
+	assert(lua_checkstack(L, bottom + 2));
+	/* Push optional arguments. */
+	while (lua_gettop(L) < bottom + 2)
+		lua_pushnil(L);
+	/*
+	 * (nil, callable) is OK, deletes the trigger
+	 * (callable, nil), is OK, adds the trigger
+	 * (callable, callable) is OK, replaces the trigger
+	 * no arguments is OK, lists all trigger
+	 * anything else is error.
+	 */
+	bool ok = true;
+	/* Name must be a string if it is passed. */
+	ok = ok && (lua_isnil(L, bottom + 2) || luaL_isnull(L, bottom + 2) ||
+		    lua_type(L, bottom + 2) == LUA_TSTRING);
+	ok = ok && (lua_isnil(L, bottom + 1) || luaL_isnull(L, bottom + 1) ||
+		    luaL_iscallable(L, bottom + 1));
+	ok = ok && (lua_isnil(L, bottom) || luaL_isnull(L, bottom) ||
+		    luaL_iscallable(L, bottom));
+	if (!ok)
+		luaL_error(L, "trigger reset: incorrect arguments");
+}
+
+/**
+ * Sets or deletes trigger by name depending on passed arguments. Value at
+ * name_idx must be a string, value at func_idx must be a callable object,
+ * nil or box.NULL. Otherwise, an error will be thrown.
+ */
+static int
+luaT_event_reset_trigger_by_name(struct lua_State *L, struct event *event,
+				 int name_idx, int func_idx)
+{
+	if (lua_type(L, name_idx) != LUA_TSTRING)
+		luaL_error(L, "name must be a string");
+	const char *trigger_name = lua_tostring(L, name_idx);
+	if (luaL_iscallable(L, func_idx)) {
+		struct func_adapter *func =
+			func_adapter_lua_create(L, func_idx);
+		event_reset_trigger(event, trigger_name, func);
+		lua_pushvalue(L, func_idx);
+		return 1;
+	} else if (lua_isnil(L, func_idx) || luaL_isnull(L, func_idx)) {
+		event_reset_trigger(event, trigger_name, NULL);
+		return 0;
+	}
+	return luaL_error(L, "func must be a callable object or nil");
+}
+
+int
+luaT_event_reset_trigger(struct lua_State *L, int bottom, struct event *event)
+{
+	assert(L != NULL);
+	assert(bottom >= 1);
+	assert(event != NULL);
+	/* Use key-value API if the first argument is a non-callable table. */
+	if (lua_gettop(L) == bottom && lua_istable(L, -1) &&
+	    !luaL_iscallable(L, -1)) {
+		lua_getfield(L, bottom, "name");
+		lua_getfield(L, bottom, "func");
+		return luaT_event_reset_trigger_by_name(L, event, -2, -1);
+	}
+	/* Old way with name support. */
+	luaT_event_reset_trigger_check_positional_input(L, bottom);
+	const int top = bottom + 2;
+	if (!lua_isnil(L, top) && !luaL_isnull(L, top))
+		return luaT_event_reset_trigger_by_name(L, event, top, bottom);
+	/*
+	 * Name is not passed - old API support.
+	 * 1. If triggers are not passed, return table of triggers.
+	 * 2. If new_trigger is passed and old_trigger is not - set
+	 * new_trigger using string representation of function as name.
+	 * 3. If old_trigger is passed and new_trigger is not - delete
+	 * trigger with by function string representation as a name.
+	 * 4. If both triggers are provided - replace old trigger with
+	 * new one if they have the same name, delete old trigger and
+	 * insert new one at the beginning of the trigger list otherwise.
```

Comment:
It's not the same behavior as the legacy one. AFAIU, if both old and new functions are specified, the old function should be replaced with the new one in the trigger list.

Reply:
The event API supports only set/delete by name, so we would have to extend the API to be able to replace a trigger with another name. And, since we are going to deprecate this interface, we decided that it's not worth it.